0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,82 @@
+0. The sender quotes to the receiver and sends a transfer with the outer request in the data field of the ILP packet. The fulfillment of this payment is the HMAC of the ILP packet using the payment key as the key.
+0. The receiver gets notified of an incoming prepared transfer with an ILP packet.
+0. The receiver generates the payment key from the outer request contained in the data field of the ILP packet, and uses it to decrypt the inner request.
+0. The receiver verifies that the amount in the incoming transfer matches the amount in the ILP packet. The receiver may also call a backend to make sure the incoming funds are expected.
+0. The receiver fulfills the incoming transfer with the HMAC of the ILP packet, using the payment key as the key.
+
+## Data Format
+
+The PSK details are divided into two request objects in the style of [HTTP Requests](https://tools.ietf.org/html/rfc7230#section-3.2). The data of the outer request is the encrypted inner request.
+
+Each of the requests starts with `PSK/1.0` and then `PUBLIC` or `PRIVATE`.  The requests are made of a list of headers, followed by data. For simplicity, duplicate headers are not allowed. A header is a UTF-8 string matching `^(.+?): (.+)$`. The first matching group is the name, and the second is the value.
+
+The headers of the outer request are visible to all parties transmitting the request, so they SHOULD contain only the bare minimum required for the intended parties to decrypt and validate the inner request.
+
+### Outer Request
+
+The outer request is formatted as below:
+
+```http
+PSK/1.0 PUBLIC
+Key-Id: fpwpAhlN588
+Key-Algorithm: HMAC-SHA-256
+Unsafe-Header: data_everyone_can_see
+
+...
+```
+
+Everything in this request is readable by all connectors between the sender and receiver. PSK payments use an [HMAC](https://en.wikipedia.org/wiki/Hash-based_message_authentication_code) of the PSK details as a fulfillment, in order to prove that the contents are unmodified. The pre-shared secret key is the key used in the HMAC.
+
+The data of this request is an encrypted binary blob. This encrypted blob contains the inner (PRIVATE) PSK request.
+
+Despite the fact that connectors can read the outer request's data, connectors should not be required to do so, nor should they rely on PSK details being present on all transfers. This information is end-to-end.
+
+The only headers in the outer request should be the ones required for the receiver to decrypt the inner request. In the above example, the decryption key for the outer request's body is: `HMAC-SHA-256(data: "fpwpAhlN588", key: secret)`, where `secret` is the pre-shared secret key.
+
+| Header | Value |
+|--------|-------|
+| `Key-Id` | _(Required)_ Base64url string used in the generation of a unique key for this payment |
+| `Key-Algorithm` | _(Required)_ The algorithm by which the Key-Id and pre-shared secret key generate the payment key |
```

Comment:
I don't understand why that would go into every payment. Wouldn't it be communicated at the time of key exchange? What do you do differently if you get a payment with an expired key?